### DISCUSSION TITLE
database migration

### DISCUSSION THREAD

**Discussion by caullla** [fluxcd/flux2#2324] [Q&A] [answered]
I'm new to fluxcd and Kubernetes.
I already have an automatic rollout. But before rolling out the new image I need to run database migration.
How do I do it?

**Comment by souleb** [comment, marked as answer]
You could use a kubernetes job to run the migration + an initContainer to wait a successful job completion. You could use https://www.google.com/url?sa=t&source=web&rct=j&url=https://github.com/groundnuty/k8s-wait-for&ved=2ahUKEwiesdqVxsP1AhXCy4UKHaipA3IQFnoECAUQAQ&usg=AOvVaw32DrY1hYr6Xhr60d_9U-NZ.

**Reply by bliednov** [reply]
@souleb Thanks for the answer! I have a follow-up question if you do not mind. What about the execution of the job in this case? Should it be run manually or is there a "FluxCD way" to automate the job execution?

P.S. I've seen ArgoCD has [resource hooks](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/) for this particular case.

**Reply by souleb** [reply]
It's a kubernetes job. Flux will reconcile it like any other kubernetes object. Put it in your path and use a flux kustomization to tell flux how to reconcile it. You will have the release yamls in the path as well (plain yamls or one of the flux source crds). 

Ps:Some people use helm hooks it seems to do the DB migration.

**Reply by Thiritin** [reply]
Hey, I'm facing this issue aswell. The big issue you will face is that a job in kubernetes is immuteable. That means once it's created you cannot change it's spec. You can set a ttl for the job, this will do the job delete itself after X seconds, this will cause the job to be recreated by FluxCD every 10 minutes.

My temporary solution, although not excellent  is to use the image updater to update the job name for the migration.
![image](https://user-images.githubusercontent.com/6755282/150669098-7ff9ee88-d201-4d6d-9502-3f11474a377b.png)

**Reply by texasbobs** [reply]
> Hey, I'm facing this issue aswell. The big issue you will face is that a job in kubernetes is immuteable. That means once it's created you cannot change it's spec. 

I believe if you use the force parameter on the kustomization, it will take care of issues related to immutable fields.

**Reply by onedr0p** [reply]
I am thinking that something like the Resource Hooks in ArgoCD would be beneficial.

There's going to be a situation where:
- You don't control the development of a helm chart, it's maintained outside your org and you would like to run a `Job` before upgrading the Helm chart (e.g. snapshot a persistent volume.)
- In regards to an `initContainer`, `k8s-wait-for` and a `Job`, you do not want to run an `Job` on a restart of the pod, only an Helm chart upgrade. There's no way I can think of how that would know the difference between a `HelmRelease` update or a restart of a `Pod`.

**Comment by bliednov** [comment]
@Thiritin proposes a good way to solve it, but I do not like the name of the job as a version. It could have some conflict if your infrastructure needs two jobs that depend on the service version.

Image automation should be able to replace the values with some pattern. It would solve the issue.
E.g. something like that
```
apiVersion: batch/v1
kind: Job
metadata:
  name: jobname- # { "$imagepolicy": "flux-system:identity:tag", "op": "append" }
...
```
In the end, the same job name should be used in `initContainers` for `k8s-wait-for`.

There is also another way to solve the issue. Using kustomize `nameSuffix`, but in this case, [k8s-wait-for](https://github.com/groundnuty/k8s-wait-for) should have an additional suffix argument for the job name. This solution is overall a bit more ugly due to additional boilerplate and custom k8s-wait-for.

So fluxcd forces me to use helm at the moment.
IMHO this use case is so standard that FluxCD should solve it without forcing devs to use helm.

**Reply by souleb** [reply]
see https://github.com/fluxcd/kustomize-controller/issues/553

**Reply by stefanprodan** [reply]
There is no need to change the Job name, set `force: true` in the Kustomization that applies jobs, and they will be recreated when an immutable field changes. Docs here https://fluxcd.io/docs/components/kustomize/kustomization/#reconciliation
